In [3]:
pip install PyPDF2


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.9/63.9 kB 2.3 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for PyPDF4: filename=PyPDF4-1.27.0-py3-none-any.whl size=61226 sha256=576a283a32cdc79479780206fb8e0651fb5dd25fc19f49c23a2fe0b260719e86
  Stored in directory: /home/axat/.cache/pip/wheels/ec/90/9c/f1815a3c733075db80497b42b70cc33f620ee7e3076bffbf77
Successfully built PyPDF4
Note: you may need to restart the kernel to use updated packages.


In [11]:
import PyPDF2

def read_pdf(file_path):
    with open(file_path, 'rb') as file:
        pdf_reader = PyPDF2.PdfReader(file)
        num_pages = len(pdf_reader.pages)
        text = ''
        for page_num in range(num_pages):
            page = pdf_reader.pages[page_num]
            text += page.extract_text()
    return text

# Example usage
file_path = '/home/axat/personal/pySpark/data/miri_tech_website_competitors_analysis_detailed.pdf'
pdf_text = read_pdf(file_path)

# Replace 'sumit' with 'robin' in the text
pdf_text = pdf_text.replace('Sumit', 'robin')
# pdf_text = pdf_text.replace('\n\n', '')

# Save the modified text as a new text file
with open('/home/axat/personal/pySpark/output/2.txt', 'w') as file:
    file.write(pdf_text)

# Print the modified text
print(pdf_text)


MiriTechWebsiteandCompetitors
Analysis
This document provides a comprehensive analysis of the MiriTech website, along with a
comparative study of its key competitors in the industry. The aim is to understand
MiriTech's online presence and how it stands in relation to its market competitors.
1.OverviewofMiriTech
MiriTech, accessible at http://miritech.com, is a technology-focused company offering
[describe services and products]. This section delves into the website's layout, content, user
experience, and unique offerings.
1.1WebsiteLayoutandDesign
An analysis of the website's design, navigation, accessibility, and overall user interface. Key
aspects include homepage layout, menu structures, and mobile responsiveness.
Placeholder for a website screenshot.
1.2ContentAnalysis
Evaluation of the quality, relevance, and variety of content available on MiriTech's website.
This includes blogs, case studies, product descriptions, and multimedia elements.
2.CompetitorsAnalysis
A detailed compari

Works as expected converting the pdf to text

In [6]:
pip install --upgrade PyMuPDF


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.4/4.4 MB 1.4 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.6/30.6 MB 2.0 MB/s eta 0:00:0000:0100:01
Note: you may need to restart the kernel to use updated packages.


In [1]:
import fitz  # PyMuPDF
import pytesseract
from PIL import Image, ImageDraw, ImageFont
import io

# Initialize Tesseract and PyMuPDF
#pytesseract.pytesseract.tesseract_cmd = r"path_to_tesseract.exe"  # Update the path to Tesseract if needed

# Function to replace text in an image
def replace_text_in_image(image, replacements, font_path='arial.ttf', font_size=12):
    """
    Replaces text in an image with new text.

    :param image: The image to modify.
    :param replacements: A list of text replacements. Each replacement is a tuple of the form
                         (old_text, (x, y, w, h), new_text), where (x, y, w, h) is the position of the
                         old text in the image.
    :param font_path: The path to the font file to use for the new text.
    :param font_size: The font size of the new text.
    :return: The modified image.
    """
    draw = ImageDraw.Draw(image)
    font = ImageFont.truetype(font_path, font_size)
    for text, position, new_text in replacements:
        draw.rectangle(position, fill='white')  # Draw rectangle to cover the old text
        draw.text((position[0], position[1]), new_text, fill='black', font=font)  # Draw new text

    return image

# Function to perform OCR and get text with positions
def ocr_image(image):
    """
    Performs OCR on an image using Tesseract and returns the recognized text along with the positions
    of each text in the image.

    :param image: The image to perform OCR on.
    :return: A list of tuples, where each tuple contains the recognized text and its position
             (x, y, width, height) in the image.
    """
    ocr_result = pytesseract.image_to_data(image, output_type=pytesseract.Output.DICT)
    text_list = ocr_result['text']
    positions = [(ocr_result['left'][i], ocr_result['top'][i], ocr_result['width'][i], ocr_result['height'][i]) for i in range(len(text_list))]
    return zip(text_list, positions)

# Load PDF
pdf_document = fitz.open('/home/axat/personal/pySpark/data/EthDev.pdf')
# new_pdf = fitz.open()

# Iterate over each page in the PDF
for page_number in range(len(pdf_document)):
    # Load the page and get the pixmap
    page = pdf_document.loadPage(page_number)
    pix = page.get_pixmap()
    img_data = pix.tobytes("png")
    image = Image.open(io.BytesIO(img_data))

    # Perform OCR on the image
    ocr_results = ocr_image(image)

    # Define text replacements - ('old_text', (x, y, w, h), 'new_text')
    # replacements = [(text, pos, 'new_text') for text, pos in ocr_results if text == 'old_text']

    # # Replace text in the image
    # new_image = replace_text_in_image(image, replacements)

    # # Convert the PIL image back to PyMuPDF pixmap
    # img_bytes = io.BytesIO()
    # new_image.save(img_bytes, format='PNG')
    # img_bytes.seek(0)
    # pix = fitz.Pixmap(fitz.csRGB, img_bytes.read())

    # # Create a new PDF page with the modified image
    # pdf_page = new_pdf.new_page(width=pix.width, height=pix.height)
    # pdf_page.insert_image(pdf_page.rect, pixmap=pix)

    #print (ocr_results)
    print (ocr_results)

# Close the original PDF
pdf_document.close()

# Save the modified PDF
#new_pdf.save('modified_pdf_file.pdf')
# new_pdf.close()

AttributeError: 'Document' object has no attribute 'loadPage'

In [2]:
import fitz  # PyMuPDF
import pytesseract
from PIL import Image, ImageDraw, ImageFont
import io

# Function to perform OCR and get text with positions
def ocr_image(image):
    """
    Performs OCR on an image using Tesseract and returns the recognized text along with the positions
    of each text in the image.

    :param image: The image to perform OCR on.
    :return: A list of tuples, where each tuple contains the recognized text and its position
             (x, y, width, height) in the image.
    """
    ocr_result = pytesseract.image_to_data(image, output_type=pytesseract.Output.DICT)
    text_list = ocr_result['text']
    positions = [(ocr_result['left'][i], ocr_result['top'][i], ocr_result['width'][i], ocr_result['height'][i]) for i in range(len(text_list))]
    return zip(text_list, positions)

# Load PDF
pdf_document = fitz.open('/home/axat/personal/pySpark/data/EthDev.pdf')

# Iterate over each page in the PDF
for page_number in range(pdf_document.page_count):
    # Load the page and get the pixmap
    page = pdf_document.load_page(page_number)
    pix = page.get_pixmap()
    img_data = pix.tobytes("png")
    image = Image.open(io.BytesIO(img_data))

    # Perform OCR on the image
    ocr_results = ocr_image(image)

    # Print OCR results
    print(f"Page {page_number + 1} OCR Results:")
    for text, position in ocr_results:
        print(f"Text: {text}, Position: {position}")

# Close the original PDF
pdf_document.close()


Page 1 OCR Results:
Text: , Position: (0, 0, 596, 842)
Text: , Position: (1, 3, 262, 11)
Text: , Position: (1, 3, 262, 11)
Text: , Position: (1, 3, 262, 11)
Text: Ethereum, Position: (1, 3, 46, 8)
Text: development, Position: (50, 3, 62, 11)
Text: documentation, Position: (115, 3, 72, 8)
Text: |, Position: (192, 4, 1, 9)
Text: ethereum.org., Position: (199, 3, 64, 11)
Text: , Position: (1, 831, 24, 8)
Text: , Position: (1, 831, 24, 8)
Text: , Position: (1, 831, 24, 8)
Text: Lof8, Position: (1, 831, 24, 8)
Text: , Position: (36, 59, 328, 17)
Text: , Position: (36, 59, 328, 17)
Text: , Position: (36, 59, 328, 17)
Text: ©, Position: (36, 62, 4, 4)
Text: @evelopers, Position: (48, 59, 82, 17)
Text: v, Position: (122, 49, 14, 31)
Text: Enterprise, Position: (144, 59, 68, 11)
Text: v, Position: (207, 49, 12, 31)
Text: Community, Position: (226, 59, 75, 11)
Text: v, Position: (295, 49, 11, 31)
Text: Q, Position: (315, 59, 10, 9)
Text: [k, Position: (359, 60, 5, 7)
Text: , Position: (261, 109,

In [16]:
pip install PyPDF2 --upgrade


Note: you may need to restart the kernel to use updated packages.


In [8]:
from PyPDF2 import PdfReader, PdfWriter


def replace_string_in_pdf(input_file, output_file, old_string, new_string):
    """
    Replaces a specific string with another string in a PDF file.

    Args:
        input_file: Path to the input PDF file.
        output_file: Path to the output PDF file.
        old_string: The string to be replaced.
        new_string: The string to replace with.
    """
    reader = PdfReader(input_file)
    writer = PdfWriter()
    new_page = writer.add_blank_page()

    for page_number in range(len(reader.pages)):
        # Extract the page content
        page_content = reader.pages[page_number].extract_text()

        # Replace the old string with the new string
        new_content = page_content.replace(old_string, new_string)

        # Create a new page object with the replaced content
        new_page = writer.add_page()
        # new_page.merge_page(reader.pages[page_number])
        # new_page.add_text(new_content)
        

    # Save the modified PDF file
    with open(output_file, "wb") as f:
        writer.write(f)


# Example usage
input_file = "/home/axat/personal/pySpark/data/miri_tech_website_competitors_analysis_detailed.pdf"
output_file = "/home/axat/personal/pySpark/output/modified_file.pdf"
old_string = "Sumit-wadhwa@india.com"
new_string = "robin-saini@india.com"

replace_string_in_pdf(input_file, output_file, old_string, new_string)
print(f"PDF file modified successfully. Output file: {output_file}")


TypeError: PdfWriter.add_page() missing 1 required positional argument: 'page'

In [9]:
pip install pdfplumber


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 16.5 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.6/45.6 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.7/118.7 kB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.0/49.0 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 22.5 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.4/4.4 MB 13.0 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 464.8/464.8 kB 10.0 MB/s eta 0:00:0000:01
Note: you may need to restart the kernel to use updated packages.


In [9]:
pip install reportlab PyPDF2


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 13.5 MB/s eta 0:00:00a 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter
from io import BytesIO
from PyPDF2 import PdfReader, PdfWriter

def replace_string_in_pdf(input_pdf_path, output_pdf_path, old_string, new_string):
    packet = BytesIO()
    # create a new PDF with ReportLab
    can = canvas.Canvas(packet, pagesize=letter)
    can.drawString(100, 100, new_string)
    can.save()

    # move to the beginning of the StringIO buffer
    packet.seek(0)
    new_pdf = PdfReader(packet)

    # open the existing PDF
    existing_pdf = PdfReader(open(input_pdf_path, "rb"))
    output = PdfWriter()

    # add the "watermark" (which is the new pdf) on the existing page
    for i in range(len(existing_pdf.pages)):
        page = existing_pdf.pages[i]
        page.merge_page(new_pdf.pages[0])
        output.add_page(page)

    # write to the output file
    with open(output_pdf_path, "wb") as outputStream:
        output.write(outputStream)

# Example usage
input_file = "/home/axat/personal/pySpark/data/miri_tech_website_competitors_analysis_detailed.pdf"
output_file = "/home/axat/personal/pySpark/output/modified_file.pdf"
old_string = "Sumit-wadhwa@india.com"
new_string = "robin-saini@india.com"

replace_string_in_pdf(input_file, output_file, old_string, new_string)

print("PDF modification complete.")


PDF modification complete.


In [6]:
from PyPDF2 import PdfReader, PdfWriter

def replace_string_in_pdf(input_pdf_path, output_pdf_path, old_string, new_string):
    # Open the existing PDF
    existing_pdf = PdfReader(input_pdf_path)

    # Create a PDF writer
    pdf_writer = PdfWriter()

    # Iterate through each page in the PDF
    for page_number in range(len(existing_pdf.pages)):
        page = existing_pdf.pages[page_number]

        # Extract text from the page
        text = page.extract_text()

        # Replace the old string with the new string
        modified_text = text.replace(old_string, new_string)

        # Create a new page with the modified text
        modified_page = pdf_writer.add_page(page)
        modified_page.merge_page(page)

        # Add modified text to the page
        modified_page.add_text(modified_text, 10, 10)  # You may need to adjust the coordinates

    # Write the modified PDF to the output file
    with open(output_pdf_path, 'wb') as output_file:
        pdf_writer.write(output_file)

# Example usage
input_file = "/home/axat/personal/pySpark/data/miri_tech_website_competitors_analysis_detailed.pdf"
output_file = "/home/axat/personal/pySpark/output/modified_file.pdf"
old_string = "Sumit-wadhwa@india.com"
new_string = "robin-saini@india.com"

replace_string_in_pdf(input_file, output_file, old_string, new_string)

print("PDF modification complete.")


AttributeError: 'PageObject' object has no attribute 'add_text'

In [9]:
import PyPDF2

def replace_text_in_pdf(input_pdf_path, output_pdf_path, text_to_replace, replacement_text):
  """
  Replaces all occurrences of "text_to_replace" with "replacement_text" in each page of the input PDF and saves the modified PDF to "output_pdf_path".

  Args:
    input_pdf_path: The path to the input PDF file.
    output_pdf_path: The path to save the modified PDF file.
    text_to_replace: The text to be replaced.
    replacement_text: The replacement text.
  """

  # Open the input PDF file.
  with open(input_pdf_path, "rb") as f:
    reader = PyPDF2.PdfReader(f)

  # Create a new PDF writer object.
  writer = PyPDF2.PdfWriter()

  # Loop through each page in the input PDF.
  for page in reader.pages:
    # Extract the page content.
    page_content = page.extractText()

    # Replace the text.
    new_page_content = page_content.replace(text_to_replace, replacement_text)

    # Create a new page object and set its content.
    new_page = PyPDF2.PageObject.createBlankPage()
    new_page.drawText(new_page_content)

    # Add the new page to the writer object.
    writer.addPage(new_page)

  # Save the modified PDF.
  with open(output_pdf_path, "wb") as f:
    writer.write(f)

# Example usage
input_pdf_path = "/home/axat/personal/pySpark/data/miri_tech_website_competitors_analysis_detailed.pdf"
output_pdf_path = "/home/axat/personal/pySpark/output/modified_file.pdf"
text_to_replace = "Sumit-wadhwa@india.com"
replacement_text = "robin-saini@india.com"

replace_text_in_pdf(input_pdf_path, output_pdf_path, text_to_replace, replacement_text)

print("Successfully replaced text in PDF.")


ValueError: seek of closed file

In [3]:
import PyPDF2

def read_pdf(file_path):
    with open(file_path, 'rb') as file:
        pdf_reader = PyPDF2.PdfReader(file)
        num_pages = len(pdf_reader.pages)
        text = ''
        pageObjeList = {}
        for page_num in range(num_pages):
            page = pdf_reader.pages[page_num]
            pageObjeList[page_num] = page.get_object()
    return pageObjeList



def read_page_object(file_path):
    with open(file_path, 'rb') as file:
        pdf_reader = PyPDF2.PdfReader(file)
        pageObj={}
        page = pdf_reader.pages[0]
      #  pageObj = page.get_object()
    return page

# Example usage
file_path = '/home/axat/personal/pySpark/data/miri_tech_website_competitors_analysis_detailed.pdf'
pdf_text = read_pdf(file_path)

obj=read_page_object(file_path)
print(obj)

# # Replace 'sumit' with 'robin' in the text
# pdf_text = pdf_text.replace('Sumit', 'robin')
# # pdf_text = pdf_text.replace('\n\n', '')

# # Save the modified text as a new text file
# with open('/home/axat/personal/pySpark/output/2.txt', 'w') as file:
#     file.write(pdf_text)

# # Print the modified text
# print(pdf_text)


{'/Contents': [IndirectObject(5, 0, 139837089888400)], '/MediaBox': [0, 0, 612, 792], '/Parent': IndirectObject(2, 0, 139837089888400), '/Resources': {'/ExtGState': {'/E1': IndirectObject(16, 0, 139837089888400)}, '/Font': {'/F1': IndirectObject(7, 0, 139837089888400), '/F2': IndirectObject(17, 0, 139837089888400), '/F3': IndirectObject(26, 0, 139837089888400)}, '/ProcSet': ['/PDF', '/Text', '/ImageB', '/ImageC', '/ImageI']}, '/Type': '/Page'}


{'/Contents': [IndirectObject(5, 0, 139837092779664)], '/MediaBox': [0, 0, 612, 792], '/Parent': IndirectObject(2, 0, 139837092779664), '/Resources': {'/ExtGState': {'/E1': IndirectObject(16, 0, 139837092779664)}, '/Font': {'/F1': IndirectObject(7, 0, 139837092779664), '/F2': IndirectObject(17, 0, 139837092779664), '/F3': IndirectObject(26, 0, 139837092779664)}, '/ProcSet': ['/PDF', '/Text', '/ImageB', '/ImageC', '/ImageI']}, '/Type': '/Page'}


import fitz  # PyMuPDF
import re
# ... import other libraries as needed ...

# Regular expressions for sensitive data
# ... define your regex patterns ...

def mask_sensitive_data(page, text_instances):
    for inst in text_instances:
        # Create a rectangle to cover the sensitive text
        rect = fitz.Rect(inst[:4])
        # You can adjust the color and opacity as needed
        page.draw_rect(rect, color=(0, 0, 0), fill=(0, 0, 0), overlay=True)

def process_pdf(file_path):
    doc = fitz.open(file_path)
    new_doc = fitz.open()

    for page_num, page in enumerate(doc):
        text = page.get_text("text")
        # Find instances of sensitive data
        sensitive_data_instances = re.finditer(your_regex_pattern, text)
        
        for match in sensitive_data_instances:
            # Mask each instance
            mask_sensitive_data(page, page.search_for(match.group()))

        new_page = new_doc.new_page(width=page.rect.width, height=page.rect.height)
        new_page.show_pdf_page(page.rect, doc, page_num)

    new_file_path = file_path.replace('.pdf', '_masked.pdf')
    new_doc.save(new_file_path)
    new_doc.close()
    doc.close()
    print(f"Masked PDF saved as {new_file_path}")

file_path = 'path_to_your_pdf_file.pdf'
process_pdf(file_path)


In [ ]:
import fitz  # PyMuPDF
import re
# ... import other libraries as needed ...

# Regular expressions for sensitive data
# ... define your regex patterns ...

def mask_sensitive_data(page, text_instances):
    for inst in text_instances:
        # Create a rectangle to cover the sensitive text
        rect = fitz.Rect(inst[:4])
        # You can adjust the color and opacity as needed
        page.draw_rect(rect, color=(0, 0, 0), fill=(0, 0, 0), overlay=True)

def process_pdf(file_path):
    doc = fitz.open(file_path)
    new_doc = fitz.open()

    for page_num, page in enumerate(doc):
        text = page.get_text("text")
        # Find instances of sensitive data
        sensitive_data_instances = re.finditer(your_regex_pattern, text)
        
        for match in sensitive_data_instances:
            # Mask each instance
            mask_sensitive_data(page, page.search_for(match.group()))

        new_page = new_doc.new_page(width=page.rect.width, height=page.rect.height)
        new_page.show_pdf_page(page.rect, doc, page_num)

    new_file_path = file_path.replace('.pdf', '_masked.pdf')
    new_doc.save(new_file_path)
    new_doc.close()
    doc.close()
    print(f"Masked PDF saved as {new_file_path}")

file_path = 'path_to_your_pdf_file.pdf'
process_pdf(file_path)


In [17]:
import fitz
from urllib.parse import urlparse
import re

# Define a function to check if a string is a valid URL
def is_url(s):
    try:
        result = urlparse(s)
        return all([result.scheme, result.netloc])
    except ValueError:
        return False

def process_pdf(file_path):
    doc = fitz.open(file_path)
    new_doc = fitz.open()

    for page_num, page in enumerate(doc):
        text = page.get_text("text")

        # Split text into words and check each for a valid URL
        words = re.findall(r'\S+', text)
        for word in words:
            if is_url(word):
                # Find the coordinates of the word on the page
                word_instances = [match.span() for match in re.finditer(re.escape(word), text)]
                for match in word_instances:
                    start, end = match
                    # Create a rectangle to cover the word
                    rect = fitz.Rect(start, match[0], end, match[1])
                    # You can adjust the color and opacity as needed
                    page.draw_rect(rect, color=(0, 0, 0), fill=(0, 0, 0), overlay=True)
                    print(f"URL Found: {word}")

        new_page = new_doc.new_page(width=page.rect.width, height=page.rect.height)
        new_page.show_pdf_page(page.rect, doc, page_num)

    new_file_path = file_path.replace('.pdf', '_masked.pdf')
    new_doc.save(new_file_path)
    new_doc.close()
    doc.close()
    print(f"Masked PDF saved as {new_file_path}")

file_path = '/home/axat/personal/pySpark/data/blockout.pdf'
process_pdf(file_path)


URL Found: https://chat.openai.com.
URL Found: https://chat.openai.com.
URL Found: https://chat.openai.com.
URL Found: https://chat.openai.com.
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https://chat.openai.com
URL Found: https

In [23]:
import fitz
import re

# Define a regex pattern for identifying URLs
url_pattern = re.compile(r'https?://\S+|www\.\S+')

def mask_sensitive_data(page, text):
    for match in url_pattern.finditer(text):
        start, end = match.span()
        
        # Get the coordinates of the matched text
        quad = page.search_for(match.group())

        if quad:
            # Extract coordinates of the first quad
            x0, y0, x1, y1 = quad[0]
        else:
            # Use coordinates based on start and end positions
            x0, y0, x1, y1 = start, 0, end, page.rect.height

        # Create a rectangle around the matched URL
        rect = fitz.Rect(x0, y0, x1, y1)
        # You can adjust the color and opacity as needed
        page.draw_rect(rect, color=(0, 0, 0), fill=(0, 0, 0), overlay=True)

def process_pdf(file_path):
    doc = fitz.open(file_path)
    new_doc = fitz.open()

    for page_num in range(doc.page_count):
        page = doc[page_num]
        text = page.get_text()
        # Mask potential URLs on the page
        mask_sensitive_data(page, text)

        new_page = new_doc.new_page(width=page.rect.width, height=page.rect.height)
        new_page.show_pdf_page(page.rect, doc, page_num)

    new_file_path = file_path.replace('.pdf', '_masked.pdf')
    new_doc.save(new_file_path)
    new_doc.close()
    doc.close()
    print(f"Masked PDF saved as {new_file_path}")

file_path = '/home/axat/personal/pySpark/data/blockout.pdf'
process_pdf(file_path)


Masked PDF saved as /home/axat/personal/pySpark/data/blockout_masked.pdf


In [24]:
def process_pdf(file_path):
    doc = fitz.open(file_path)

    for page_num in range(doc.page_count):
        page = doc[page_num]
        text = page.get_text()
        
        # Print the text content to the console
        print(f"Page {page_num + 1} Text:\n{text}")

    doc.close()

file_path = '/home/axat/personal/pySpark/data/blockout.pdf'
process_pdf(file_path)


Page 1 Text:
Header3
1
Lorem ipsum
·Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Nunc ac faucibus odio.
Vestibulum neque massa, scelerisque sit amet ligula eu, congue molestie mi. Praesent ut
varius sem. Nullam at porttitor arcu, nec lacinia nisi. Ut ac dolor vitae odio interdum
condimentum.
Vivamus
dapibus
sodales
ex,
vitae
malesuada
ipsum
cursus
convallis. Maecenas sed egestas nulla, ac condimentum orci. Mauris diam felis,
vulputate ac suscipit et, iaculis non est. Curabitur semper arcu ac ligula semper, nec luctus
nisl blandit. Integer lacinia ante ac libero lobortis imperdiet. Nullam mollis convallis ipsum,
ac accumsan nunc vehicula vitae. Nulla eget justo in felis tristique https://chat.openai.com.
Morbi sit amet tortor quis risus auctor condimentum. Morbi in ullamcorper elit. Nulla iaculis
tellus sit amet mauris tempus https://chat.openai.com.
Maecenas mauris lectus, lobortis et purus mattis, blandit dictum tellus.
· Maecenas non lorem quis tellus placerat varius.
· N

MASKING FOR LINKS WORKING

In [26]:
import fitz
import re

# Define a regex pattern for identifying URLs
url_pattern = re.compile(
    r'https?://[^\s/$.?#].[^\s]*',
    re.IGNORECASE
)

def mask_sensitive_data(page, text_instances):
    for inst in text_instances:
        # Create a rectangle to cover the sensitive text
        rect = fitz.Rect(inst[:4])
        # You can adjust the color and opacity as needed
        page.draw_rect(rect, color=(0, 0, 0), fill=(0, 0, 0), overlay=True)

def process_pdf(file_path):
    doc = fitz.open(file_path)
    new_doc = fitz.open()

    for page_num in range(doc.page_count):
        page = doc[page_num]
        text = page.get_text()

        # Find instances of sensitive data (URLs in this case)
        sensitive_data_instances = re.finditer(url_pattern, text)

        for match in sensitive_data_instances:
            # Mask each instance
            mask_sensitive_data(page, page.search_for(match.group()))
            
            # Print the matched URL to the console
            print(f"Page {page_num + 1}, Matched URL: {match.group()}")

        # Add the current page to the new document
        new_page = new_doc.new_page(width=page.rect.width, height=page.rect.height)
        new_page.show_pdf_page(page.rect, doc, page_num)

    new_file_path = file_path.replace('.pdf', '_masked.pdf')
    new_doc.save(new_file_path)
    new_doc.close()
    doc.close()

    print(f"Masked PDF saved as {new_file_path}")

file_path = '/home/axat/personal/pySpark/data/blockout.pdf'
process_pdf(file_path)


Page 1, Matched URL: https://chat.openai.com.
Page 1, Matched URL: https://chat.openai.com.
Page 1, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com.
Page 3, Matched URL: https://chat.openai.com.
Page 3, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 5

works for grey and links 

In [36]:
import fitz
import re

# Define a regex pattern for identifying URLs
url_pattern = re.compile(
    r'https?://[^\s/$.?#].[^\s]*',
    re.IGNORECASE
)

def mask_sensitive_data(page, text_instances):
    for inst in text_instances:
        # Create a rectangle to cover the sensitive text
        rect = fitz.Rect(inst[:4])
        # You can adjust the color and opacity as needed
        page.draw_rect(rect, color=(0.99, 0.99, 0.99), fill=(0.05, 0.05, 0.05, 0.05), overlay=True)
def process_pdf(file_path):
    doc = fitz.open(file_path)
    new_doc = fitz.open()

    for page_num in range(doc.page_count):
        page = doc[page_num]
        text = page.get_text()

        # Find instances of sensitive data (URLs in this case)
        sensitive_data_instances = re.finditer(url_pattern, text)

        for match in sensitive_data_instances:
            # Mask each instance
            mask_sensitive_data(page, page.search_for(match.group()))
            
            # Print the matched URL to the console
            print(f"Page {page_num + 1}, Matched URL: {match.group()}")

        # Add the current page to the new document
        new_page = new_doc.new_page(width=page.rect.width, height=page.rect.height)
        new_page.show_pdf_page(page.rect, doc, page_num)

    new_file_path = file_path.replace('.pdf', '_masked.pdf')
    new_doc.save(new_file_path)
    new_doc.close()
    doc.close()

    print(f"Masked PDF saved as {new_file_path}")

file_path = '/home/axat/personal/pySpark/data/blockout.pdf'
process_pdf(file_path)


Page 1, Matched URL: https://chat.openai.com.
Page 1, Matched URL: https://chat.openai.com.
Page 1, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 2, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com
Page 3, Matched URL: https://chat.openai.com.
Page 3, Matched URL: https://chat.openai.com.
Page 3, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 4, Matched URL: https://chat.openai.com
Page 5